### ISSUE TITLE
MS SQL Vulnerability assessment times out when using access/SAS key

### ISSUE THREAD

**Issue by almir0** [hashicorp/terraform-provider-azurerm#15425]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.15.4 on windows_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.87.0

### Affected Resource(s)

* `azurerm_mssql_server_vulnerability_assessment`

### Terraform Configuration Files

```hcl
resource "azurerm_mssql_server_security_alert_policy" "server" {
  resource_group_name = var.resource_group_name
  server_name         = var.server_name
  state               = var.mssql_server_vulnerability_assessment.server_vulnerability_assesment.server_security_alert_policy_state
}

resource "azurerm_mssql_server_vulnerability_assessment" "server" {
  server_security_alert_policy_id = azurerm_mssql_server_security_alert_policy.server.id
  storage_container_path          = "${data.azurerm_storage_account.vulnerability_assessment.primary_blob_endpoint}vulnerability-assessment/"
  storage_account_access_key      = data.azurerm_storage_account.vulnerability_assessment.primary_access_key

  recurring_scans {
    enabled                   = var.mssql_server_vulnerability_assessment.server_vulnerability_assesment.reccuring_scans_enabled
    email_subscription_admins = var.mssql_server_vulnerability_assessment.server_vulnerability_assesment.email_subscription_admins
    emails                    = var.mssql_server_vulnerability_assessment.server_vulnerability_assesment.emails
  }
  depends_on = [
    azurerm_mssql_server_security_alert_policy.server
  ]
}
```

### Expected Behaviour

Vulnerability assessment for Azure SQL server should be created with recurring scans and emails to subscription owner plus additional emails.

### Actual Behaviour

After approx 15 minutes the apply proccess times out with Error: updataing mssql server vulnerability assessment: sql.ServerVulnerabilityAssessmentsClient#CreateOrUpdate: Failure responding to request: StatusCode=504 — Original Error: autorest/azure: Service returned an error. Status=504 Code=”GatewayTimeout” Message=”The gateway did not receive a response from ‘Microsoft.Sql’ within the specified time period.”

### Steps to Reproduce

1. `terraform apply`  configuration above.

### References

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/mssql_server_vulnerability_assessment suggests following:

[storage_account_access_key](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/mssql_server_vulnerability_assessment#storage_account_access_key) - (Optional) Specifies the identifier key of the storage account for vulnerability assessment scan results. If [storage_container_sas_key](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/mssql_server_vulnerability_assessment#storage_container_sas_key) isn't specified, storage_account_access_key is required.

storage_container_sas_key - (Optional) A shared access signature (SAS Key) that has write access to the blob container specified in storage_container_path parameter. If storage_account_access_key isn't specified, storage_container_sas_key is required.

If storage account access key and/or sas key is removed and SQL managed identity is used that has Storage Blob Data Contributor on storage account then the terraform apply runs successfully.

That should be added to the documentation.

* #0000

**Comment by sinbai** [comment]
@almir0 thanks for opening this issue. I could not reproduce this issue when running `terraform apply` with the following tf config. Could you repro by the tf config below? And could you try to check what might be wrong with the steps compared to yours? Also, if you could provide the complete step-by-step repro steps and tf config, it would be help me to repro/troubleshooting. 

```
provider "azurerm" {
  features {}
}

terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "=2.87.0"
    }
  }
}

resource "azurerm_resource_group" "test" {
  name     = "rg_example"
  location = "eastus"
}

resource "azurerm_mssql_server" "test" {
  name                         = "sqlserverexample"
  resource_group_name          = azurerm_resource_group.test.name
  location                     = azurerm_resource_group.test.location
  version                      = "12.0"
  administrator_login          = "mradministrator"
  administrator_login_password = "thisIsDog11"
}

resource "azurerm_storage_account" "test" {
  name                     = "accsaexample"
  resource_group_name      = azurerm_resource_group.test.name
  location                 = "eastus"
  account_tier             = "Standard"
  account_replication_type = "GRS"
}

resource "azurerm_storage_container" "test" {
  name                  = "acctestsc-example"
  storage_account_name  = azurerm_storage_account.test.name
  container_access_type = "private"
}

resource "azurerm_mssql_server_security_alert_policy" "server" {
  resource_group_name = azurerm_resource_group.test.name
  server_name         = azurerm_mssql_server.test.name
  state               = "Enabled"
}

resource "azurerm_mssql_server_vulnerability_assessment" "server" {
  server_security_alert_policy_id = azurerm_mssql_server_security_alert_policy.server.id
  storage_container_path          = "${azurerm_storage_account.test.primary_blob_endpoint}${azurerm_storage_container.test.name}/"
  storage_account_access_key      = azurerm_storage_account.test.primary_access_key

  recurring_scans {
    enabled                   = true
    email_subscription_admins = true
    emails = [
      "email@example1.com",
      "email@example2.com"
    ]
  }
  depends_on = [
    azurerm_mssql_server_security_alert_policy.server
  ]
}
```

**Comment by almir0** [comment]
Thanks for quick response!

I have 3 environments that are using same codebase, 2 (dev/test) of them are located in one Azure subscription and the third (production) have it's own subscription. So the code I pasted do work as intended in dev/test but not in production.

The main difference between the dev/test and prod is that in production Azure Defender for cloud plan is enabled on subscription level for all Azure SQL Database servers.

This is the order we created all this in.

* SQL database server is created using terraform when Defender for cloud were disabled.
* Azure Defender for cloud is enabled on subscription level
* The resource azurerm_mssql_server_vulnerability_assessment is added to existing code that were used for creating SQL databases.

Try to reproduce in those three steps and I'll try your code in my production subscription as soon as possible.

**Comment by almir0** [comment]
Unfortunately I was not able to reproduce this error simply by using the code you provided. My production environment is more complex than that, we are using private endpoints for storage account, outbound firewall rules for SQL server as well as NSGs for subnets when it comes to SQL database.

All this works fine when I do it via Azure portal and by removing the `storage_account_access_key      = azurerm_storage_account.test.primary_access_key` line the vulnerability assessment is forced to use SQL system assigned managed identity to connect to the storage account.

I'm totally fine with that since I prefer to use managed identity before access keys, it's just that by reading the documentation I got understanding that I must use either access key och sas key and I was surprised that it just worked without that line and that the behaviour is not documented.

**Comment by sinbai** [comment]
@almir0 thanks for your feedback. Re "_it's just that by reading the documentation I got understanding that I must use either access key och sas key and I was surprised that it just worked without that line and that the behaviour is not documented._", I submitted a PR #15465 to add comments for the storage_account_access_key and storage_container_sas_key in tf doc. It has been merged and would be released this Friday. 

Based on the comments (storage_account_access_key and storage_container_sas_key only apply if the storage account is not behind a virtual network or a firewall) added in my PR #15465, if your use-case applies this case, I assume that's why it does not work. 

Per the above conclusion, are you okey if we close this issue?
